test/integration/tcp_proxy_odcds_integration_test.cc

```diff
@@ -0,0 +1,259 @@
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}, odcds_stream_));
+
+  // This upstream is listening on the endpoint of ``new_cluster``. It starts to serve tcp_proxy.
+  FakeRawConnectionPtr fake_upstream_connection;
+  ASSERT_TRUE(fake_upstreams_.back()->waitForRawConnection(fake_upstream_connection));
+
+  ASSERT_TRUE(fake_upstream_connection->write("hello"));
+  tcp_client->waitForData("hello");
+
+  ASSERT_TRUE(tcp_client->write("world"));
+  ASSERT_TRUE(fake_upstream_connection->waitForData(5));
+
+  ASSERT_TRUE(fake_upstream_connection->close());
+  ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
+  tcp_client->waitForHalfClose();
+  tcp_client->close();
+}
+
+// Verify only one delta xds response is needed for multiple tcp_proxy requests.
+TEST_P(TcpProxyOdcdsIntegrationTest, RepeatedRequest) {
+  expected_upstream_connections_ = 10;
+  initialize();
+
+  // Establish tcp connections to the target Envoy.
+  for (auto n = expected_upstream_connections_; n != 0; n--) {
+    tcp_clients_.push_back(makeTcpConnection(lookupPort("tcp_proxy")));
+  }
+
+  // The delta CDS stream is established.
+  auto result = fake_upstreams_.front()->waitForHttpConnection(*dispatcher_, xds_connection_);
+  RELEASE_ASSERT(result, result.message());
+  result = xds_connection_->waitForNewStream(*dispatcher_, odcds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  odcds_stream_->startGrpcStream();
+
+  // Verify the delta CDS request and respond without providing the cluster.
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {"new_cluster"}, {},
+                                           odcds_stream_));
+  sendDeltaDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+      Config::TypeUrl::get().Cluster, {new_cluster_}, {}, "1", odcds_stream_);
+  EXPECT_TRUE(compareDeltaDiscoveryRequest(Config::TypeUrl::get().Cluster, {}, {}, odcds_stream_));
+
+  // This upstream is listening on the endpoint of ``new_cluster``.
+  for (auto n = expected_upstream_connections_; n != 0; n--) {
+    fake_upstream_connections_.push_back(nullptr);
+    auto& fake_upstream_connection = fake_upstream_connections_.back();
+    ASSERT_TRUE(fake_upstreams_.back()->waitForRawConnection(fake_upstream_connection));
+    ASSERT_TRUE(fake_upstream_connection->write("hello"));
+  }
+
+  for (auto& tcp_client : tcp_clients_) {
+    tcp_client->waitForData("hello");
+    ASSERT_TRUE(tcp_client->write("world"));
+  }
+
+  for (auto& fake_upstream_connection : fake_upstream_connections_) {
+    ASSERT_TRUE(fake_upstream_connection->waitForData(5));
+    ASSERT_TRUE(fake_upstream_connection->close());
+    ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
+  }
+
+  for (auto& tcp_client : tcp_clients_) {
+    tcp_client->waitForHalfClose();
+    tcp_client->close();
+  }
+}
+
+TEST_P(TcpProxyOdcdsIntegrationTest, ShutdownConnectionOnTimeout) {
```

Comment:
Yeah, this timer is not a local timer. Instead each cluster maintains a single timer instance.
If there was a pending timer which will be fired in 1 second (which could be created by another tcp proxy request)
Even a new tcp_proxy instance set the timer to 5 seconds,
the new tcp_proxy will be notified in 1 second rather than 5. 
I can add a test case here.

Is this behavior acceptable? If not, we can set another tcp_proxy maintained timer when the ondemand timeout is 
implemented. API: https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/filters/network/tcp_proxy/v3/tcp_proxy.proto#L104

Reply:
When I add the counter I realize any xds response won't trigger the `timeout`. xds response can only cancel the timer if the response offers the requested cluster or deleted the requested cluster.
The cluster lookup time timeout is triggered at desired future wall time. Since this timer is invisible at tcp_proxy, it's not realistic to cancel the timer, tcp_proxy can only removed itself from the notification list. This case is tested in the "tcp connection close" test case.

What I prefer is to define the look up timeout behavior in the follow up. Along with the question that if a per tcp connection timer is required.

Reply:
so prior to this PR there was an assumption that upstream was unhealthy, we'd error out and close connections after a connection duration timeout.  I think as is, we could have an unexpected issue where connections can simply accumulate. I don't think we should unhide the config without either adding a timer or adding an warning of this failure mode but I'm happy to get a second opinion from Greg or Matt if you'd prefer?

Reply:
I am not sure if I get it. The previous connection_duration_timer_ is respected. We don't really have more downstream connection in this PR with assumption that  the connection_duration_timer_

If tcp_proxy doesn't set connection_duration_timer_, each connection is expected to live no more than 5 seconds if on demand is enabled and the cluster is not ready. I don't expect tcp connection will accumulate.

Regarding the tcp request failing mode, The new connection failure mode in this PR is the cluster lookup failure. It's user opt-in. Tcp proxy owner must specify the OD setting. Also, it's up to the control plane to push all the clusters as in the past(http server push :) ).

Reply:
Sorry, I meant the connection establishment timeout.  Basically before we'd close connections if we couldn't talk to upstream, and now we may or may not.

cc @mattklein123 @ggreenway for a second opinion on if this new path merits a timeout.

Reply:
@ggreenway @alyssawilk What if we end this PR with a runtime config, then start another thread on the expect timer behavior? hitchhike timer in ODCDS system. 

My perception is that a configurable value is more important than the timer behavior. Yeah, if you are ok with the timer mechanism, I can inline the PR timer API
```
message OnDemand {
   ...
    // The timeout for on demand cluster lookup. If the CDS cannot return the required cluster,
    // the downstream request will be closed with the error code detail NO_CLUSTER_FOUND.
    // [#not-implemented-hide:]
    google.protobuf.Duration timeout = 3;
  }
```

Reply:
yeah that works for me

Reply:
I create https://github.com/envoyproxy/envoy/issues/20633 as follow up. Can we use the default 60s now and land this PR?